[1단계 - 장바구니] 루나(오신정) 미션 제출합니다.

ShinjungOh opened the pull request:
# 🛒 장바구니

안녕하세요 해먼드!!
지난 영화리뷰 미션에 이어 또 만나뵙게 되어 반갑습니다 🙌
장바구니 1단계 미션 완료해서 리뷰 요청 드립니다. 이번에도 잘 부탁드릴게요! 🙇‍♀️

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] RTL 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포 링크 기입: [🔗 배포 사이트](https://shinjungoh.github.io/react-shopping-cart/)
- [x] 리뷰어가 장바구니 추가를 쉽게 할 수 있도록 Curl 명령어를 DM 으로 전달해주세요.

### 리뷰 요청 & 논의하고 싶은 내용

### 테스트 통과

![CleanShot 2025-05-29 at 17 25 54@2x](https://github.com/user-attachments/assets/188682e6-b284-4696-bd32-b8b0a969763c)

### 1) 상태 설계 의도

상태 구조는 다음과 같아요.

```
// 실제 상태 (2개만 관리)
- cartItems: 서버 데이터 (DataContext)
- checkedItems: 선택된 ID 배열 (useCartSelection)

// 파생 값 (계산으로 처리)
- isAllChecked = checkedItems.length === cartItems.length
- price = 선택된 아이템들의 합계
- shippingFee = price < 100000 ? 3000 : 0
```

프로젝트에서 `DataContext` 하나로 상품 데이터와 기타 상태를 관리하고 있어요. 다만, 상품 선택 기능(전체 선택, 개별 선택 등)은 재사용성을 고려해 별도의 `useCartSelection` 훅으로 분리했어요.

이렇게 설계한 이유는 다음과 같아요.
- DataContext: 서버와의 데이터 동기화에 집중 (fetching, caching, refetch)
- useCartSelection: UI 상태 관리에 집중 (선택/해제 로직)

두 관심사를 분리해서 각각의 역할이 명확해지고, 추후 선택 기능을 다른 컴포넌트에서도 쉽게 재사용할 수 있도록 했어요. 
최상위 컴포넌트에서 이 훅의 반환 값을 props로 하위 컴포넌트에 전달하는 방식으로 단방향 데이터 흐름을 유지했어요.



### 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

이전까지는 작은 컴포넌트부터 구현하는 바텀업 방식으로 개발했어요. 작은 단위부터 만드는 것이 편할 것이라 생각했지만, 실제로는 여러 어려움을 겪었어요. 상태 관리의 복잡성, 파생 상태 때문에 상태 동기화 문제 등이 있었어요.

그래서 이번 미션은 탑다운 방식으로 개발을 진행했어요. 

 1. 단일 진실 공급원(Single Source of Truth)
  - 장바구니 선택 상태를 useCartSelection 커스텀 훅에서 중앙 관리
  - 모든 하위 컴포넌트는 props로 상태를 전달받아 사용
  - 상태 동기화 문제가 자연스럽게 해결됨

2. 파생 상태 최소화
  - 가격 계산, 배송비 등은 checkedItems로부터 직접 계산
  - 불필요한 state 생성을 피하고 순수 함수로 처리
  - 리렌더링 최적화와 코드 간결성 확보

탑다운 방식으로 전환하면서 먼저 전체 화면을 그리고, 추후에 분리하는 방식으로 리팩토링 했습니다. 
특히 상태 관리가 필요한 복잡한 UI에서는 데이터 흐름을 먼저 설계하고 구현하는 것이 훨씬 효율적이었어요. 

페어와 함께 본격적인 개발에 들어가기 전에 상태 및 컴포넌트에 대한 설계를 확실하게 하고 간 것이 많이 도움되었어요!

### 3) 이번 리뷰를 통해 논의하고 싶은 부분

프로젝트에서 `DataContext` 하나로 상품 데이터와 기타 상태를 관리하고 있어요. 
다만, 상품 선택 기능(전체 선택, 개별 선택 등)은 재사용성을 고려해 별도의 `useCartSelection` 훅으로 분리하여 이 훅의 반환 값을 props 형태로 자식 컴포넌트에 전달하고 있어요. 

하지만 이 방식으로 구현하다 보니 몇 가지 고민이 생겼어요.

- Props drilling으로 인한 중간 컴포넌트의 불필요한 props 전달
- 컴포넌트 간 의존성 증가로 인한 유지보수 복잡도 상승

`useCartSelection` 훅과 기존 `DataContext`를 조합해 장바구니 관련 모든 상태를 한 곳에서 관리하거나, 컨텍스트 구조를 개선하는 등 추가로 고려할 수 있는 리팩토링 방안이 있을지 해먼드의 의견이 궁금합니다~!

```
 const { checkedItems, setCheckedItems, isAllChecked, handleAllCheck } =
    useCartSelection(cartItems);
```

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 클라이언트 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. MSW/Test

- 주요 기능을 적절히 정의하였나요?
- 주요 기능/예외에 대한 테스트가 충분히 이루어졌나요?

woowa-euijinkk committed:
chore: 의존성 제거

aydenote committed:
chore: emotion, react-router 의존성 추가

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
docs: README 기능 목록 작성

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: prettier 설정

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 전역 스타일 적용

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 레이아웃 스타일 적용

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 헤더 스타일 적용

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
refactor: 헤더 컴포넌트 title 타입 옵셔널 추가

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 주문확인 페이지 UI 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 버튼 높이 수정

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 장바구니 상품 컴포넌트 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 장바구니 페이지 UI 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: 정보 icon 이미지 추가

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 장바구니 상품 목록 UI 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 장바구니 상품 목록 스크롤 적용

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 장바구니 상품 선택 체크 박스 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: 불필요한 주석 제거

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: API 통신 관련 env 작성

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 메인 페이지에서 데이터 패칭 기능 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 장바구니 상품 정보 화면에 렌더링 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 장바구니 상품 수량 변경 기능 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
docs: 장바구니 상품 수량 변경 기능 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 장바구니 상품 삭제 기능 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: DataContext 생성

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
refactor: Context API 상태 관리 방법 개선

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
refactor: 장바구니 API 호출 함수 관심사 분리

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
docs: 구현 사항 업데이트

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: 불필요한 콘솔 제거

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 선택된 상품 가격 계산 기능 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 주문확인 페이지 동적 UI 구현

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 주문 금액 boderline 스타일 변경

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
refactor: 장바구니 상품 선택 관련 로직 커스텀 훅으로 분리

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
feat: 페이지 하단 버튼 컴포넌트 생성 및 적용

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
refactor: 관심사에 따라 컴포넌트 분리

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

woowa-euijinkk committed:
docs: PR template

aydenote committed:
docs: PR 템플릿 추가

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
design: 분리된 스타일 코드 컴포넌트 내부로 이동, 공통 스타일은 파일로 분리

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
refactor: 메인 페이지 컴포넌트로 관심사 분리

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: 배포 관련 설정

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: vitest 의존성 설치

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
chore: vitest 환경 설정

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
fix: 수량 감소 로직 조건문 수정

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
test: CartItem의 API 호출 및 UI 상호작용 테스트

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
test: CartPage의 비즈니스 로직 및 라우팅 테스트

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
test: 커스텀 훅 테스트 작성

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

aydenote committed:
test: 라우트 테스트 오류 수정

Co-authored-by: Luna <ShinjungOh@users.noreply.github.com>

ShinjungOh committed:
chore: 배포 홈페이지 주소 변경

JunilHwang commented on src/types/cart.ts:
```diff
@@ -0,0 +1,18 @@
+  product: Product;
+}
+
+export interface CartInfo {
+  id: number;
+  quantity: number;
+}
```
CartInfo 는 지금 안 쓰이고 있네요!

ShinjungOh replied:
사용하지 않는 CartInfo 타입 제거했어요!

[remove: 불필요 타입 제거](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/470d48a6f5f06ffc725e85ba99c9bb4c6684dfa8)

JunilHwang commented on test/CartItem.test.tsx:
```diff
@@ -0,0 +1,96 @@
+  },
+}));
+
+describe('CartItem', () => {
```
지금 통합 테스트만 작성 되었는데요,
핵심 로직이나 공통 로직을 유틸 함수와 훅으로 분리한 다음에 단위 테스트도 같이 작성해주시면 좋을 것 같아요!
그래야 코드의 수정에 대해 빠르게 테스트를 통해 피드백 할 수 있답니다 ㅎㅎ

ShinjungOh replied:
넵!! 우선 분리할 수 있는 로직을 훅과 유틸함수로 분리한 다음, 테스트 코드를 작성하도록 할게요ㅎㅎ

JunilHwang commented on test/CartItem.test.tsx:
```diff
@@ -0,0 +1,96 @@
+      </div>
+    );
+  },
+}));
```
SelectBox를 목킹하는 이유가 있을까요!? 그대로 사용하면 어떤 문제가 있는걸까요?

ShinjungOh replied:
SelectBox를 mocking하면 `SelectBox`의 내부 구현 로직과 독립적으로 테스트를 할 수 있어요. 
원래는 `CartItem` 컴포넌트의 동작만 테스트하려는 의도였어요.

`SelectBox`를 모킹하지 않는다면 SelectBox와 CartItem 컴포넌트가 결합되어 실제로 유저가 사용하는 것처럼 검증할 수 있다는 장점이 있을 것 같아요.
SelectBox는 단순한 UI 컴포넌트이기 때문에 모킹하지 않고 실제로 사용하는 것이 나을 것 같네요!!

모킹하지 않고 실제 컴포넌트를 사용하는 것으로 개선했습니다~!

[test: CartItem 테스트에서 SelectBox 모킹 제거](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/cd6e53062bbf25e707fad94e0b39bc31bb62b555)

JunilHwang commented on test/CartItem.test.tsx:
```diff
@@ -0,0 +1,96 @@
+
+    await waitFor(() => {
+      expect(cartApi.patchIncreaseQuantity).toHaveBeenCalledWith(mockCartItem);
+    });
```
waitFor 대신 act를 이용해서 표현해보시면 좋겠어요..!

waitFor 을 쓰면 안 되는 이유도 같이 고민해보시면 좋겠습니다.

ShinjungOh replied:
말씀하신대로 waitFor와 act의 차이에 대해 학습해봤어요.

### waitFor vs act의 차이

- waitFor: 조건이 충족될 때까지 폴링하며 기다림
	- 외부 상태나 DOM 변화를 기다릴 때 사용
- act: React의 상태 업데이트를 동기적으로 처리하도록 보장

### 현재 코드에서 waitFor를 사용할 경우

단순히 기다리는 것이 목적인지, 상태 변화를 유발하는 것이 목적인지 테스트 의도를 명확하게 드러내지 못한다는 점, 
React의 동작 원리와 불일치한다는 점 (배치 업데이트나 이펙트 실행을 제대로 시뮬레이션하지 못할 수 있음)
=> waitFor보다 act가 적합

### act가 더 적절한 이유

act는 모든 업데이트가 DOM에 반영되도록 리액트의 렌더링 사이클을 보장하고,
유저 상호작용이나 상태 변화를 명시적으로 래핑

-> act는 사용자 인터랙션, 상태 업데이트 등 동기적 작업을 테스트할 때 사용하는 것이 적절
현재 저의 코드에서는 리액트 상태 변화의 결과로 API 호출이 되고 있기 때문에 `act`가 더 적합한 것이 맞네요..! 


[test: waitFor를 act로 교체 및 테스트 통과하도록 코드 수정](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/e75773bf53107da933d9243368ee9e41365fa257)

JunilHwang commented on src/hooks/useCartSelection.ts:
```diff
@@ -0,0 +1,50 @@
+      const allIds = cartItems.content.map((item: CartProduct) => item.id);
+      setCheckedItems(allIds);
+    }
+  };
```
handle 이라고 표현하기엔 이벤트를 다루고 있지 않아서,
`checkAll` 이라고 표현해주면 어떨까요!?
이 함수가 하는 일을 이름으로 표현하는거죠 ㅎㅎ

ShinjungOh replied:
넵 해당 함수명을 `checkAll`로 교체했습니다!

[refactor: handleAllCheck 함수명을 checkAll로 변경](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/fb8f5a595d306dd5c4bd7f3b7ef83f5bd535fa5a)

JunilHwang commented on src/hooks/useCartSelection.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+export const useCartSelection = (cartItems: CartItemsResponse | undefined) => {
+  const [checkedItems, setCheckedItems] = useState<number[]>([]);
+
+  useEffect(() => {
+    if (cartItems?.content && cartItems.content.length > 0) {
+      const itemIds = cartItems.content.map((item: CartProduct) => item.id);
+      setCheckedItems(itemIds);
+    } else {
+      setCheckedItems([]);
+    }
+  }, [cartItems]);
```
useEffect 대신에 useMemo 로 표현할 수 있지 않을까요?

```tsx
const checkedItems = useMemo(() => cartItems.content.map((item: CartProduct) => item.id). [cartItems]);
```
요로코롬...

JunilHwang commented on src/pages/CartPage.tsx:
```diff
@@ -0,0 +1,104 @@
+        {cartItems.content.length > 0 ? (
+          <>
+            <CartMain
+              isAllChecked={isAllChecked}
+              checkedItems={checkedItems}
+              setCheckedItems={setCheckedItems}
+              price={price}
+              shippingFee={shippingFee}
+              totalPrice={totalPrice}
+              handleAllCheck={handleAllCheck}
```
props를 부모에서 정의한 이름을 그대로 자식에게 넘겨주고 있는데요,
자식 컴포넌트 입장에서의 이름을 고려해주면 좋겠어요..!

JunilHwang replied:
가령, handleAllCheck 가 아니라 onAllCheckClick 이라고 표현해야 자연스럽지 않나요!?

ShinjungOh replied:
아래로 쭉 남겨주신 코멘트를 보면서 개선한 결과, `onChange={checkAll}`라고 넘겨주는 것으로 수정했습니다!

[refactor: props 네이밍 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/37efffef11480a2881feabf4c34064ca1f531d20)

JunilHwang commented on src/components/Cart/CartMain.tsx:
```diff
@@ -0,0 +1,54 @@
+}: CartMainProps) {
+  return (
+    <>
+      <SelectAllBox isAllChecked={isAllChecked} handleAllCheck={handleAllCheck} />
```
SelectAllBox 입장에서는 handleAllCheck 가 아니라 onChange 혹은 onClick 이 되겠죠?

ShinjungOh replied:
`onChange`로 수정했습니다! 

[refactor: props 네이밍 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/37efffef11480a2881feabf4c34064ca1f531d20)

JunilHwang commented on src/components/SelectBox/SelectAllBox.tsx:
```diff
@@ -0,0 +1,40 @@
+        <CheckboxContainer>
+          <HiddenCheckbox
+            type="checkbox"
+            checked={isAllChecked}
```
native html tag 들은 boolean에 is를 붙이지 않는답니다 ㅎㅎ
그래서 여기서도 checked 라고 표현되고 있어요.

SelectAllBox도 isAllChecked 가 아니라 checked 로 받아와야 이 컴포넌트 입장에서 자연스럽지 않을까요!?

ShinjungOh replied:
그러고 보니 `disabled`도 is를 붙이지 않고 사용하는군요! 
`checked`로 받도록 이름을 개선했어요ㅎㅎ

[refactor: props 네이밍 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/37efffef11480a2881feabf4c34064ca1f531d20)

JunilHwang commented on src/components/SelectBox/SelectAllBox.tsx:
```diff
@@ -0,0 +1,40 @@
+interface SelectBoxProps {
+  isAllChecked: boolean;
+  handleAllCheck: (checked: boolean) => void;
+}
```
결국 부모에서 정의한 이름이 자식까지 계속 전파되는 모습인데요,
그렇다면 props가 아니라 차라리 context 를 이용해서 표현하는게 좋지 않나 싶어요..!

ShinjungOh replied:
props drilling의 문제를 props의 이름으로 겪고 있네요..!
`context`를 사용하는 방법도 생각해 보도록 할게요👀

[refactor: props 네이밍 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/37efffef11480a2881feabf4c34064ca1f531d20)

JunilHwang commented on src/components/SelectBox/SelectBox.tsx:
```diff
@@ -0,0 +1,48 @@
+  onRemove: () => void;
+}
+
+function SelectBox({ cartItem, checkedItems, setCheckedItems, onRemove }: SelectBoxProps) {
```
select 입장에서는 onRemove 라는 행위를 하지 않아요.
onRemove가 내포하는 뜻이 정확히 무엇일까요?

ShinjungOh replied:
`SelectBox` 컴포넌트에서 `onRemove`는 삭제 버튼을 클릭했을 때 실행되는 함수이고, 
 SelectBox 입장에서는 삭제 버튼의 클릭 이벤트를 처리한다는 의미예요.

제가 사용한 `onRemove`는 삭제한다는 행위 자체를 의미하는 이름으로 느껴지네요.
 
 영어라 잘 와닿지 않아서 AI와 질답을 해봤는데 다음과 같이 알려줬어요.
 
 ⏺ onRemove와 onDeleteClick의 차이점:

**onRemove**
  - "제거될 때" 실행되는 콜백
  - 제거라는 결과/행위에 초점
  - 어떤 이벤트로 인해 제거가 일어나는지는 명시하지 않음
  - 더 추상적인 표현

**onDeleteClick**
  - "삭제 버튼을 클릭했을 때" 실행되는 콜백
  - 삭제 버튼의 클릭 이벤트에 초점
  - 사용자의 구체적인 인터랙션을 명시
  - 더 구체적인 표현


컴포넌트의 실제 동작을 더 정확하게 표현하기 위해 `onDeleteClick`으로 수정했어요.


[refactor: SelectBox의 onRemove를 onDeleteClick로 네이밍 변경](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/002cc8af74bf879fdcc838b8956c45334242a3ff)

JunilHwang commented on src/components/Cart/CartItem.tsx:
```diff
@@ -0,0 +1,168 @@
+    name: 'cartItems',
+  });
+
+  const handleRemoveCartItem = async () => {
+    await removeCartItem(cartItem);
+    refetch();
+  };
+
+  const handleIncreaseQuantity = async () => {
+    await patchIncreaseQuantity(cartItem);
+    refetch();
+  };
+
+  const handleDecreaseQuantity = async () => {
+    if (cartItem.quantity === 1) {
+      handleRemoveCartItem();
+      return;
+    }
+
+    await patchDecreaseQuantity(cartItem);
+    refetch();
+  };
```
여기에도 handle을 붙일 이유가 없는 것 같아요..!
이 함수가 하는 일을 이름으로 표현해주세요~

ShinjungOh replied:
`handle`을 제거하는 방향으로 함수명을 개선했어요!

[refactor: CartItem 이벤트 핸들러 함수명 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/b8befd4655feb9abce9d3546632c2d3047e4418d)

JunilHwang commented on src/components/Cart/CartItem.tsx:
```diff
@@ -0,0 +1,168 @@
+}
+
+function CartItem({ cartItem, checkedItems, setCheckedItems }: CartItemProps) {
+  const { refetch } = useData({
+    fetcher: getCartItems,
+    name: 'cartItems',
+  });
+
+  const handleRemoveCartItem = async () => {
+    await removeCartItem(cartItem);
+    refetch();
+  };
+
+  const handleIncreaseQuantity = async () => {
+    await patchIncreaseQuantity(cartItem);
+    refetch();
+  };
+
+  const handleDecreaseQuantity = async () => {
+    if (cartItem.quantity === 1) {
+      handleRemoveCartItem();
+      return;
+    }
+
+    await patchDecreaseQuantity(cartItem);
+    refetch();
+  };
+
```
여기서 쓰이는 코드들도 훅으로 묶어서 표현해주면 어떨까요?

ShinjungOh replied:
장바구니 아이템 관련 로직을 `useCartItem` 커스텀 훅으로 이동시켰어요.
`CartItem` 컴포넌트는 UI 렌더링만, 상태 관리와 API 호출은 커스텀 훅에서 처리하도록 분리해
UI와 비즈니스 로직이 구분되도록 리팩토링 했습니다~!

[refactor: CartItem의 로직을 useCartItem 커스텀 훅으로 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/490b94d5a66ecb1ea1c4dbe7bc814e91b8574eca)

JunilHwang commented on src/components/Header/Header.tsx:
```diff
@@ -0,0 +1,62 @@
+  const navigate = useNavigate();
+  const isDefault = variant === 'default';
+
+  const handlePageBack = () => {
+    navigate(-1);
+  };
```
`goToBack` 이라고 표현해주면 좋을 것 같아요!

ShinjungOh replied:
해당 함수명을 알려주신대로 변경했습니다!

[refactor: Header의 뒤로가기 기능 함수명 goToBack으로 변경](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/9b225b4bec730715395d42422229023518f9c2ca)

JunilHwang commented on src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,39 @@
+      {children}
+    </BaseButton>
+  );
+}
```
Props가 `ComponentProps<'button'>` 을 사용해주면 어떨까요!?
그래야 더 유연한 Button 으로 사용될 수 있을 것 같아요!

ShinjungOh replied:
Button 컴포넌트에 `ComponentProps<'button'>`을 추가했어요. 
기본 button HTML 속성을 사용할 수 있는 유연한 컴포넌트로 개선했어요~!

[refactor: Button 컴포넌트에서 ComponentProps 적용해 유연한 컴포넌트로 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/7c0b91cbb3b932f8ac27a787528e5b7e76d01037)

JunilHwang commented on src/apis/cart.ts:
```diff
@@ -0,0 +1,64 @@
+import { CartProduct } from '../types/cart';
+import { BASE_URL, USER_TOKEN } from './env';
+
+export const getCartItems = async () => {
+  try {
+    const response = await fetch(`${BASE_URL}/cart-items?page=0&size=20`, {
+      headers: {
+        Authorization: `Basic ${USER_TOKEN}`,
+        'content-type': 'application/json',
+      },
+    });
+    return await response.json();
+  } catch (error) {
+    throw new Error('장바구니 목록을 불러오는 중 에러 발생');
+  }
+};
+
+export const patchIncreaseQuantity = async (cartItem: CartProduct) => {
+  try {
+    await fetch(`${BASE_URL}/cart-items/${cartItem.id}`, {
+      method: 'PATCH',
+      headers: {
+        Authorization: `Basic ${USER_TOKEN}`,
+        'content-type': 'application/json',
+      },
+      body: JSON.stringify({
+        quantity: cartItem.quantity + 1,
+      }),
+    });
+  } catch (error) {
+    throw new Error('장바구니 상품 증가시 오류 발생');
+  }
+};
```
반복되는 코드가 많이 보이네요!
apiClient 같은걸 하나 만들어서 관리해주면 어떨까요!?

요런 느낌으로..

https://axios-http.com/docs/instance

ShinjungOh replied:
반복되던 `headers, BASE_URL` 설정을 `apiClient`에서 한 번에 관리할 수 있도록 `apiClient`를 생성했어요.
모든 API 호출에 동일한 에러 처리와 응답 처리 로직을 사용할 수 있고,
토큰이나 Base url이 변경되었을 때 한 파일만 수정하면 되도록 개선했어요!

[refactor: apiClient 생성 및 cart.ts의 반복 코드 제거](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/32ff2b5a8d4cc6fcebbc52571744128679655807)

JunilHwang commented on src/pages/CartPage.tsx:
```diff
@@ -0,0 +1,104 @@
+import CartMain from '../components/Cart/CartMain';
+
+function CartPage() {
+  const { data: cartItems } = useData({
+    fetcher: getCartItems,
+    name: 'cartItems',
+  });
+  const navigate = useNavigate();
+  const { checkedItems, setCheckedItems, isAllChecked, handleAllCheck } =
+    useCartSelection(cartItems);
+
+  const price = cartItems?.content
+    ? cartItems.content
+        .filter((item: CartProduct) => checkedItems.includes(item.id))
+        .reduce((sum: number, item: CartProduct) => sum + item.product.price * item.quantity, 0)
+    : 0;
+
+  const totalCount = cartItems?.content
+    ? cartItems.content
+        .filter((item: CartProduct) => checkedItems.includes(item.id))
+        .reduce((sum: number, item: CartProduct) => sum + item.quantity, 0)
+    : 0;
+
+  const hasItems = checkedItems.length > 0;
+  const needsShippingFee = price < 100000;
+  const shippingFee = hasItems && needsShippingFee ? 3000 : 0;
+
+  const totalPrice = price + shippingFee;
+
+  if (!cartItems) {
+    return null;
+  }
+
+  const descriptionMessage = () => {
+    if (cartItems.content.length > 0) {
+      return `현재 ${cartItems.content.length}종류의 상품이 담겨있습니다.`;
+    }
+  };
+
+  const isDisabled = checkedItems.length === 0;
```
이런 것들도 훅으로 묶어주면 어떨까요?

JunilHwang replied:
순수 함수로 표현할 수 있는건 순수함수로 분리해주면 좋겠어요!

ShinjungOh replied:
CartPage 리팩토링을 완료했어요. 리팩토링 과정을 적어볼게요!

1. `CartPage`에 있던 비즈니스 로직을 `useCart` 커스텀훅을 생성해 분리했어요.
2. useCart 훅에 있는 계산 로직을 `cartCalculations`로 분리했어요. - 순수함수 분리
3. `useCart` 커스텀훅에 너무 많은 관심사가 혼재되어 있어서, 단일 책임 원칙을 지키도록 잘게 나눴어요. 
  - useCartData: 데이터 패칭
  - useCartCalculations: 가격 및 수량 계산
  - useCartUI: UI 상태 관리

=>  이렇게 개선한 결과 `useCart`는 다른 훅들을 조합하는 역할만 수행하도록 했어요.


[refactor: CartPage 비즈니스 로직을 useCart 훅으로 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/19300844b806fc1666987a0c8bbc44f4a8c8e54a)
[refactor: useCart 훅을 관심사별로 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/4cf21457ab439765be228bebe2beb01e9ddad70e)

JunilHwang commented on src/pages/CartPage.tsx:
```diff
@@ -0,0 +1,104 @@
+    : 0;
+
+  const hasItems = checkedItems.length > 0;
+  const needsShippingFee = price < 100000;
```
100000, 3000 등은 상수로 표현할 수 있지 않을까요?

ShinjungOh replied:
배송비와 금액 관련해서 상수로 분리할 수 있는 것들을 상수화했어요.

[refactor: 주문 금액 관련 매직넘버 상수화](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/117a326379d96d3b66ea304d45a721bc59704740)

JunilHwang commented on src/hooks/useCartSelection.ts:
```diff
@@ -0,0 +1,50 @@
+    } else {
+      const allIds = cartItems.content.map((item: CartProduct) => item.id);
+      setCheckedItems(allIds);
+    }
```
```suggestion
    setCheckedItems(checked ? [] : cartItems.content.map((item: CartProduct) => item.id));
```
이렇게 표현할 수 있을 것 같아요!

ShinjungOh replied:
알려주신대로 삼항 연산자를 사용하는 방법으로 리팩토링해서 `if-else` 블록을 제거했어요.

[refactor: checkAll 로직에 삼항 연산자 적용해 간결하게 리팩토링](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/dacaac967a1696a12a774d010a0797f02f0e9b01)

JunilHwang commented on src/hooks/useCartSelection.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+export const useCartSelection = (cartItems: CartItemsResponse | undefined) => {
+  const [checkedItems, setCheckedItems] = useState<number[]>([]);
+
+  useEffect(() => {
+    if (cartItems?.content && cartItems.content.length > 0) {
+      const itemIds = cartItems.content.map((item: CartProduct) => item.id);
+      setCheckedItems(itemIds);
+    } else {
+      setCheckedItems([]);
+    }
+  }, [cartItems]);
```
`cartItems.content.map((item: CartProduct) => item.id))`를 여기저기서 사용하고 있어서, 이렇게 값으로 하나 만들어서 재활용해주면 어떨까요?

```tsx
const cartIds = cartItems.content.map((item: CartProduct) => item.id));
const cartContains = cartItems.content?.length > 0

useEffect(() => {
  setCheckedItems(cartContains ? cartIds : []);
}, [cartContains, cartIds]);
```

하단에서도 다양하게 재활용할 수 있을 것 같네요..!

ShinjungOh replied:
넵 재활용할 수 있는 코드 확인했어요! 스텝2때 반영하도록 할게요!!

JunilHwang commented on src/pages/CartPage.tsx:
```diff
@@ -0,0 +1,104 @@
+  const navigate = useNavigate();
+  const { checkedItems, setCheckedItems, isAllChecked, handleAllCheck } =
+    useCartSelection(cartItems);
+
+  const price = cartItems?.content
+    ? cartItems.content
+        .filter((item: CartProduct) => checkedItems.includes(item.id))
+        .reduce((sum: number, item: CartProduct) => sum + item.product.price * item.quantity, 0)
+    : 0;
+
+  const totalCount = cartItems?.content
+    ? cartItems.content
+        .filter((item: CartProduct) => checkedItems.includes(item.id))
+        .reduce((sum: number, item: CartProduct) => sum + item.quantity, 0)
+    : 0;
```
반복적으로 쓰이는 값들이 보이네요 ㅎㅎ 
앞선 리뷰처럼 추상화해서 사용할 수 있을 것 같아요.

ShinjungOh replied:
앞선 코멘트와 마찬가지로, 리팩토링 가능한 부분 확인했습니다~! 
마찬가지로 스텝2에 적용해볼게요💪🔥

JunilHwang commented on src/components/Cart/CartItem.tsx:
```diff
@@ -0,0 +1,168 @@
+  setCheckedItems: React.Dispatch<React.SetStateAction<number[]>>;
+}
+
+function CartItem({ cartItem, checkedItems, setCheckedItems }: CartItemProps) {
```
CartItem 입장에서는
각각
- data(혹은 value)
- checked
- onClick

이렇게 표현할 수 있겠네요!

ShinjungOh replied:
CartItem 입장에서 props의 이름을 명확히 하도록 개선했어요.

  1. CartItem은 단일 아이템에 대한 정보만 받도록 개선
  2. 체크 상태는 boolean으로 받기
  3. 체크 변경 이벤트는 콜백으로 받기

CartList와 CartItem의 역할도 명확해졌어요.
- CartList : 아이템 체크 상태 관리
- CartItem : 개별 아이템 표시 및 상호작용

[refactor: CartItem props 네이밍 개선 및 CartList와 CartItem 책임 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/96cceace9b52b8f8ea619525c67abce6b06ac875)

JunilHwang commented on src/App.tsx:
```diff
@@ -1,9 +1,25 @@
-      <h1>react-shopping-cart</h1>
+      <Global styles={GlobalStyle} />
+      <Layout>
+        <DataProvider>
```
지금은 DataProvider가 한 개 사용 중인데요,
다뤄야 하는 데이터가 많을수록 병목이 되지않을까요!?

가령, 지금은 api로 cart만 가져오고 있는데, cart 말고 다른 데이터를 가져와야한다면 어떻게 해야좋을까요!?

아니면 useSyncExternalStore 같은걸 이용해보시면 좋을 것 같아요!

ShinjungOh replied:
`useData`처럼 구현한 것은 사실 이전 <상품 목록> 미션에서의 요구사항이었는데요, 
저와 페어는 이전 미션에서 이 방식대로 구현하는 것을 제대로 소화하지 못해서 아쉬움이 있었어요. 
그래서 이번 <장바구니> 미션에서 적용해보자! 하고 시도해본 방식이예요. 

이번에는 cart 데이터만 사용하기 때문에 cart 외에 다른 데이터를 가져오는 것에 신경쓰지 못했는데,
알려주신 [useSyncExternalStore](https://ko.react.dev/reference/react/useSyncExternalStore) 방법도 학습해 보도록 하겠습니다!!ㅎㅎㅎ

JunilHwang reviewed (changes requested):
안녕하세요 루나!
장바구니 미션을 함께하게 된 리뷰어 해먼드입니다 ㅎㅎ 다시 만나서 반가워요!

---

제가 이번에 vscode로 리뷰를 했더니 ㅋㅋ 코멘트가 하나씩 달렸네요ㅠㅠ 죄송합니다..

일단 전체적인 내용을 조합해보자면, 부모 컴포넌트에서 정의된 이름을 자식 컴포넌트에서 그대로 사용하고 있어요. 이건 굉장히 좋지 못한 패턴이라고 생각합니다 ㅠㅠ

html tag를 사용하는 모습을 토대로 비교해보자면 이렇습니다.

- 보편적인 모습: `<input value={cartItem.name} />`
- 현재 루나가 정의한 모습: `<input cartItemName={cartItem.name} />`

어떤 차이인지 이해 되실까요..!?

네이밍을 하는게 어렵다면 아예 context를 활용하는 방법도 있습니다 ㅎㅎ 관심사를 하나로 묶어주는거죠.
컴포넌트를 재활용할게 아니라면 context도 좋은 대안이 되리라 생각합니다!

이 외의 내용은 상세 피드백을 확인해주시면 될 것 같아요!

---

테스트는 통과하고 있으나 커버리지가 부족한 것 같습니다.

<img width="481" alt="image" src="https://github.com/user-attachments/assets/6761bfb9-618f-491a-b756-dce32dd39f2b" />

나머지 부분은 어떻게 채울 수 있을지 고민해보시면 좋을 것 같아요!
그리고 실패에 대한 테스트도 필요하지 않을까요!?

---

> useCartSelection 훅과 기존 DataContext를 조합해 장바구니 관련 모든 상태를 한 곳에서 관리하거나, 컨텍스트 구조를 개선하는 등 추가로 고려할 수 있는 리팩토링 방안이 있을지 해먼드의 의견이 궁금합니다~!

말씀해주신 것 처럼 모든 상태를 Context 내부에서 관리하는 방법이 있을 것 같아요 ㅎㅎ
`리렌더링이 최적화` 라고 위에서 말씀해주시긴 했는데.. 문제는 리렌더링이 계속 발생하는 구조라서요.
어차피 context가 최상단에 감싸져 있고
context의 값이 변경되면 자연스럽게 렌더링이 전파됩니다.
결국 의도한 최적화는 발생하지 않는거죠..

그렇다면 아싸리 관리하기 쉽게 cart에 대한 것들을 context에 같이 묶어주면 어떨까 싶어요!
다른 대안으로는 useSyncExternalStore 를 사용하는 방법이 있답니다 ㅎㅎ
최적화에 대한 부분이라기보단, 외부 상태 관리에 대한 로직을 바깥에 정의하고, 이걸 react 시스템에 결합하는 과정이랄까..!?

ShinjungOh committed:
refactor: 주문 금액 관련 매직넘버 상수화

ShinjungOh committed:
remove: 불필요 타입 제거

ShinjungOh committed:
refactor: handleAllCheck 함수명을 checkAll로 변경

ShinjungOh committed:
refactor: props 네이밍 개선

ShinjungOh committed:
refactor: CartItem 이벤트 핸들러 함수명 개선

ShinjungOh committed:
refactor: SelectBox의 onRemove를 onDeleteClick로 네이밍 변경

ShinjungOh committed:
refactor: 주문 금액에 따른 배송비 안내 메시지의 금액을 상수로 교체

ShinjungOh committed:
refactor: Header의 뒤로가기 기능 함수명 goToBack으로 변경

ShinjungOh committed:
refactor: Button 컴포넌트에서 ComponentProps 적용해 유연한 컴포넌트로 개선

ShinjungOh committed:
refactor: CartItem의 로직을 useCartItem 커스텀 훅으로 분리

ShinjungOh committed:
refactor: apiClient 생성 및 cart.ts의 반복 코드 제거

ShinjungOh committed:
refactor: checkAll 로직에 삼항 연산자 적용해 간결하게 리팩토링

ShinjungOh committed:
refactor: CartItemsResponse 타입을 공통으로 분리

ShinjungOh committed:
fix: 빈 API 응답 처리 오류 수정

- response.json() 직접 호출 시 빈 응답에서 발생하는 파싱 오류 수정
- response.text()로 먼저 읽어 빈 응답 확인 후 안전하게 JSON 파싱

ShinjungOh committed:
refactor: CartPage 비즈니스 로직을 useCart 훅으로 분리

- useCart 훅에서 계산 로직을 cartCalculations로 분리

ShinjungOh committed:
refactor: 배송비가 없는 경우에 사용되는 매직넘버 0 상수화

ShinjungOh committed:
refactor: useCart 훅을 관심사별로 분리

ShinjungOh committed:
test: waitFor를 act로 교체 및 테스트 통과하도록 코드 수정

ShinjungOh committed:
test: CartItem 테스트에서 SelectBox 모킹 제거

ShinjungOh committed:
refactor: CartItem props 네이밍 개선 및 CartList와 CartItem 책임 분리

ShinjungOh committed:
refactor: lint, prettier 미적용 코드 수정

ShinjungOh commented:
# 🛒 장바구니 - 리팩토링

안녕하세요 해먼드~! 
피드백 반영을 마치고 리뷰 요청 드립니다!

제가 일정 관리에 어려움을 겪고 있다보니 피드백 반영이 너무 늦어졌네요ㅠ.ㅠ 죄송합니다😭
시간 관계상 스텝1에서 다 반영하지 못한 사항이 있는데(테스트 커버리지, 반복적으로 사용되는 코드를 추상화해서 재활용)
이 부분은 스텝2를 진행하면서, 리팩토링하면서 마저 채워넣도록 하겠습니다!!🔥

리팩토링 과정에서 변경된 사항을 정리해봅니다!

<br />

## ⚙️ 주요 변경 사항

### 1.  네이밍/코드 스타일 개선

함수명, 변수명, props 이름이 명확하지 못한 부분이 있었는데 이를 개선했어요.
제 프로젝트에서 계속 반복된 문제였네요🥲 
이름을 사용하는 곳(컴포넌트 등)에서 명확하고 역할을 드러낼 수 있도록 고심해야겠어요.
아니면 해당 부분은 AI에게 적극적으로 의견을 구할 수도 있겠다는 생각이 드네요ㅎㅎ

[refactor: handleAllCheck 함수명을 checkAll로 변경](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/fb8f5a595d306dd5c4bd7f3b7ef83f5bd535fa5a)
[refactor: props 네이밍 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/37efffef11480a2881feabf4c34064ca1f531d20)
[refactor: CartItem 이벤트 핸들러 함수명 개선](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/b8befd4655feb9abce9d3546632c2d3047e4418d)
[refactor: SelectBox의 onRemove를 onDeleteClick로 네이밍 변경](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/002cc8af74bf879fdcc838b8956c45334242a3ff)
[refactor: Header의 뒤로가기 기능 함수명 goToBack으로 변경](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/9b225b4bec730715395d42422229023518f9c2ca)
[refactor: lint, prettier 미적용 코드 수정](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/1fbe5d217f4a85a68ed4710d245d882265e53685)


### 2. 구조 개선 및 커스텀 훅 분리

CartPage, CartItem 등 비즈니스 로직이 들어가있는 컴포넌트에서
비즈니스 로직을 커스텀훅으로 분리하는 작업을 진행했어요. 
커스텀 훅에서도 분리할 수 있는 코드가 있다면 유틸로 분리해 관심사와 책임을 명확하게 나누고자 했어요.

[refactor: CartItem의 로직을 useCartItem 커스텀 훅으로 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/490b94d5a66ecb1ea1c4dbe7bc814e91b8574eca)
[refactor: apiClient 생성 및 cart.ts의 반복 코드 제거](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/32ff2b5a8d4cc6fcebbc52571744128679655807)
[refactor: CartPage 비즈니스 로직을 useCart 훅으로 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/19300844b806fc1666987a0c8bbc44f4a8c8e54a)
[refactor: useCart 훅을 관심사별로 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/4cf21457ab439765be228bebe2beb01e9ddad70e)
[refactor: CartItem props 네이밍 개선 및 CartList와 CartItem 책임 분리](https://github.com/woowacourse/react-shopping-cart/pull/342/commits/96cceace9b52b8f8ea619525c67abce6b06ac875)


### 3. 기타 변경사항

- Button 컴포넌트 유연성 증가
- 테스트 코드 SelectBox 모킹 제거
- 테스트가 통과하도록 코드 수정
- if/else문 대신 삼항 연산자로 코드 간결화
- 공통 타입 분리
- 배송비, 주문 수량 관련 매직넘버 상수화
- 불필요 코드 제거

JunilHwang commented on src/apis/cart.ts:
```diff
@@ -0,0 +1,46 @@
+import { CartProduct, CartItemsResponse } from '../types/cart';
+import { apiClient } from './apiClient';
+
+export const getCartItems = async (): Promise<CartItemsResponse> => {
+  try {
+    return await apiClient<CartItemsResponse>('/cart-items?page=0&size=20');
```
응답 타입은 자연스럽게 추론된답니다!

```suggestion
export const getCartItems = async () => {
  try {
    return await apiClient<CartItemsResponse>('/cart-items?page=0&size=20');
```

ShinjungOh replied:
확인했어요! 응답 타입 제거했습니다~!

[refactor: 불필요 타입 제거](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/fa12572a998c3be982883649a0c952d8a9f4538c)

JunilHwang commented on src/apis/cart.ts:
```diff
@@ -0,0 +1,46 @@
+export const getCartItems = async (): Promise<CartItemsResponse> => {
+  try {
+    return await apiClient<CartItemsResponse>('/cart-items?page=0&size=20');
+  } catch (error) {
+    throw new Error('장바구니 목록을 불러오는 중 에러 발생');
+  }
```
apiClient에서 발생한 에러 정보가 다 손실될 것 같아요..!

JunilHwang commented on src/apis/cart.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+};
+
+export const patchIncreaseQuantity = async (cartItem: CartProduct) => {
+  try {
+    return await apiClient(`/cart-items/${cartItem.id}`, {
+      method: 'PATCH',
+      body: {
+        quantity: cartItem.quantity + 1,
+      },
+    });
+  } catch (error) {
+    throw new Error('장바구니 상품 증가시 오류 발생');
+  }
+};
+
+export const patchDecreaseQuantity = async (cartItem: CartProduct) => {
+  try {
+    return await apiClient(`/cart-items/${cartItem.id}`, {
+      method: 'PATCH',
+      body: {
+        quantity: cartItem.quantity - 1,
+      },
+    });
+  } catch (error) {
+    throw new Error('장바구니 상품 감소시 오류 발생');
+  }
+};
+
+export const removeCartItem = async (cartItem: CartProduct) => {
+  try {
+    return await apiClient(`/cart-items/${cartItem.id}`, {
+      method: 'DELETE',
+    });
+  } catch (error) {
+    throw new Error('장바구니 상품 삭제시 오류 발생');
+  }
+};
```
전체적으로 요청 후에 바로 try/catch, throw를 하고 있어요.
http 요청을 실패할 때, fetch가 만들어준 기본적인 에러 정보들이 있는데 이게 다 소실되는 모습입니다.
cart를 사용하는 바깥쪽(컴포넌트 혹은 훅)에서 해주면 어떨까 싶네요!

JunilHwang commented on src/components/Cart/CartFooter.tsx:
```diff
@@ -0,0 +1,37 @@
+
+function CartFooter({ price, shippingFee, totalPrice }: CartFooterProps) {
+  const renderCartPrice = [
+    { title: '주문 금액', price, variant: 'default' as CartPriceVariant },
+    { title: '배송비', price: shippingFee, variant: 'shipping' as CartPriceVariant },
+    { title: '총 결제 금액', price: totalPrice, variant: 'total' as CartPriceVariant },
```
```suggestion
    { title: '주문 금액', price, variant: 'default' as const },
    { title: '배송비', price: shippingFee, variant: 'shipping' as const },
    { title: '총 결제 금액', price: totalPrice, variant: 'total' as const },
```
타입 오류 때문이라면, 이렇게 표현할 수도 있답니다 ㅎㅎ

ShinjungOh replied:
반영했습니다ㅎㅎ

[refactor: as 타입 단언으로 타입 에러 해결 및 불필요 함수를 값으로 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/ab23fc05ac05933763e6edb02eec419c5ab3178a)

JunilHwang commented on src/components/SelectBox/SelectBox.tsx:
```diff
@@ -0,0 +1,43 @@
+  onDeleteClick: () => void;
+}
+
+function SelectBox({ cartItem, checked, onChange, onDeleteClick }: SelectBoxProps) {
+  return (
+    <ModifyRow>
+      <CheckboxContainer>
+        <HiddenCheckbox
+          data-id={cartItem.id}
```
select box 에겐 cartItem이 아니라 id가 필요해요!
전체적인 값을 다 받아올 필요는 없어보이네요 ㅎㅎ

ShinjungOh replied:
그렇네요! 해당 컴포넌트의 props를 필요한 값인 id만 받아오도록 개선했습니다~!

[refactor: SelectBox의 props를 cartItem에서 id로 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/1602fcac8dec479adc044649d5fae09e4b79080a)

JunilHwang commented on src/context/DataContext.tsx:
```diff
@@ -0,0 +1,45 @@
+export function useData<T>({ fetcher, name }: { fetcher: () => Promise<T>; name: string }) {
+  const { data, setData } = useContext(DataContext);
+
+  const request = useCallback(() => {
+    fetcher().then((res) => {
+      setData((data) => {
+        return { ...data, [name]: res };
+      });
+    });
+  }, [fetcher, name, setData]);
```
에러에 대한 처리도 있어야 하지 않나 싶어요!

JunilHwang commented on src/hooks/useCart.ts:
```diff
@@ -0,0 +1,34 @@
+import { useCartUI } from './useCartUI';
+
+export const useCart = () => {
+  const cartItems = useCartData();
+
+  const { checkedItems, setCheckedItems, isAllChecked, checkAll } = useCartSelection(cartItems);
+
+  const { price, totalCount, shippingFee, totalPrice } = useCartCalculations({
+    items: cartItems?.content,
+    checkedIds: checkedItems,
+  });
+
+  const { descriptionMessage, isDisabled } = useCartUI({
+    itemCount: cartItems?.content?.length ?? 0,
+    checkedItemsCount: checkedItems.length,
+  });
```
이쁘게 잘 묶어주셨네요 ㅎㅎ

JunilHwang commented on src/hooks/useCartUI.ts:
```diff
@@ -0,0 +1,16 @@
+}
+
+export const useCartUI = ({ itemCount, checkedItemsCount }: UseCartUIProps) => {
+  const descriptionMessage = () => getCartDescription(itemCount);
```
함수일 필요가 있을까요?

```suggestion
  const descriptionMessage = getCartDescription(itemCount);
```

값으로 쓰여도 무방해보여요!

ShinjungOh replied:
굳이 함수로 사용하지 않아도 되는 코드였네요. 개선했어요~!

[refactor: as 타입 단언으로 타입 에러 해결 및 불필요 함수를 값으로 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/ab23fc05ac05933763e6edb02eec419c5ab3178a)

JunilHwang commented on src/utils/cartCalculations.ts:
```diff
@@ -0,0 +1,27 @@
+import { CartProduct } from '../types/cart';
+import { FREE_SHIPPING_FEE, SHIPPING_FEE, SHIPPING_FEE_THRESHOLD } from '../constants/cartConfig';
+
+export const filterCheckedItems = (items: CartProduct[], checkedIds: number[]): CartProduct[] => {
+  return items.filter((item) => checkedIds.includes(item.id));
+};
+
+export const calculateTotalPrice = (items: CartProduct[]): number => {
+  return items.reduce((sum, item) => sum + item.product.price * item.quantity, 0);
+};
+
+export const calculateTotalQuantity = (items: CartProduct[]): number => {
+  return items.reduce((sum, item) => sum + item.quantity, 0);
+};
+
+export const calculateShippingFee = (price: number, hasItems: boolean): number => {
+  const needsShippingFee = price < SHIPPING_FEE_THRESHOLD;
+  return hasItems && needsShippingFee ? SHIPPING_FEE : FREE_SHIPPING_FEE;
+};
+
+export const getCartDescription = (itemCount: number): string | undefined => {
+  if (itemCount > 0) {
+    return `현재 ${itemCount}종류의 상품이 담겨있습니다.`;
+  }
+
+  return undefined;
+};
```
이런 유틸 함수들에 대해 테스트를 작성해주면 어떨까요!?

ShinjungOh replied:
2단계에서 테스트를 작성했어요!

[test: 유틸 함수 테스트 작성](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/1d7e0ce87ea89f4bc9162f89cd0362329246715f)

JunilHwang commented on src/components/Cart/CartList.tsx:
```diff
@@ -0,0 +1,51 @@
+}
+
+function CartList({ checkedItems, setCheckedItems }: CartListProps) {
+  const { data: cartItems } = useData<CartItemsResponse>({
+    fetcher: getCartItems,
+    name: 'cartItems',
+  });
```
useCartData 로 받아오면 되지 않을까요!?

JunilHwang replied:
context는 하나인데
useData는 최상위와 여기에서 호출하고 있어요.
그러면 useEffect가 상위 레이어와 여기에서 총 두 번 호출 되겠죠?

이게 의도한게 맞는지 궁금합니다..!

JunilHwang replied:
https://github.com/user-attachments/assets/6a39fd57-9ceb-46ff-b550-771b0947f8f9

요로코롬...

JunilHwang replied:
데이터를 호출하는 로직과 데이터를 관리하는 로직이 결합되어있어서 발생하는 문제라고 볼 수 있답니다!

JunilHwang commented on src/components/Cart/CartMain.tsx:
```diff
@@ -0,0 +1,57 @@
+  onChange: (checked: boolean) => void;
+}
+
+function CartMain({
+  checked,
+  checkedItems,
+  setCheckedItems,
+  price,
+  shippingFee,
+  totalPrice,
+  onChange,
+}: CartMainProps) {
+  return (
+    <>
+      <SelectAllBox checked={checked} onChange={onChange} />
+      <CartList checkedItems={checkedItems} setCheckedItems={setCheckedItems} />
+
+      <CartInfo>
+        <InfoIconImage src={infoIcon} alt="infoIcon" />
+        <p>
+          총 주문 금액이 {SHIPPING_FEE_THRESHOLD.toLocaleString()}원 이상일 경우 무료 배송됩니다.
+        </p>
+      </CartInfo>
+
+      <CartFooter price={price} shippingFee={shippingFee} totalPrice={totalPrice} />
+    </>
+  );
+}
```
어차피 context를 통해 데이터를 받아오고 있어요.
그렇다면 최상위 계층 (CartPage) 에서 모든 값을 만들어서 내려주는게 아니라
하위 컴포넌트에서 필요한걸 원천 데이터를 통해 조합하여 사용할 수 있지 않을까요?

JunilHwang commented on src/pages/OrderConfirmPage.tsx:
```diff
@@ -0,0 +1,57 @@
+        <SubTitle>총 결제 금액</SubTitle>
+        <Title>{price.toLocaleString()}원</Title>
+      </Container>
+      <Button disabled={true}>결제하기</Button>
```
```tsx
<Button disabled>결제하기</Button>
```
이렇게 표현해도 무방해보이네요!

JunilHwang commented on test/CartPage.test.tsx:
```diff
@@ -0,0 +1,194 @@
+  },
+}));
+
+describe('CartPage', () => {
```
훅을 작성해주셔서, 훅에 대한 테스트를 추가해주시면 좋을 것 같아요..!

JunilHwang commented on test/CartPage.test.tsx:
```diff
@@ -0,0 +1,194 @@
+  it('총 가격이 올바르게 계산되어야 한다', async () => {
+    renderComponent();
+
+    await waitFor(() => {
+      expect(screen.getByTestId('price')).toHaveTextContent('130000');
+    });
```
전체적으로 waitFor도 제거해주시면 좋을 것 같아요!

ShinjungOh replied:
CartPage 테스트 코드에 있던 `waitFor`를 제거했어요!

[test: waitFor 제거 및 DataProvider 모킹으로 API 호출 에러 해결](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/cfb9c505babecc11490fb7c7b09d7a0b42d795ef)

JunilHwang reviewed (approved):
안녕하세요 루나!
리팩토링 꼼꼼하게 잘 해주셨네요 ㅎㅎ 고생하셨습니다!

일단 전체적으로 살펴봤을 때 context의 이점을 살리지 못하고 있는게 느껴져요.
결국 최상위 레이어에서 조립하여 내려주는 방식은 똑같아서, 차라리 하위 컴포넌트에서 더 적극적으로 context의 값을 참조하면 어떨까 싶습니다 ㅎㅎ

이 외에도 자잘한 피드백 남겼으니 다음 단계 진행하실 때 참고해주세요!
고생하셨습니다!
